packages/@aws-cdk/aws-cloudfront/lib/edge-function.ts

```diff
@@ -0,0 +1,209 @@
+import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
+import * as ec2 from '@aws-cdk/aws-ec2';
+import * as iam from '@aws-cdk/aws-iam';
+import * as lambda from '@aws-cdk/aws-lambda';
+import * as ssm from '@aws-cdk/aws-ssm';
+import {
+  App, BootstraplessSynthesizer, Construct, ConstructNode, DefaultStackSynthesizer, IStackSynthesizer,
+  ResourceEnvironment, Stack, StackProps, Token,
+} from '@aws-cdk/core';
+import * as resources from '@aws-cdk/custom-resources';
+
+/** Properties for creating a Lambda@Edge function */
+export interface EdgeFunctionProps extends lambda.FunctionProps { }
+
```

Comment:
`lambda.Version` already has an `edgeArn()` API  that lets customers create an edge function from `Function` construct - https://github.com/aws/aws-cdk/blob/e349004a522e2123c1e93bd3402dd7c3f9c5c17c/packages/%40aws-cdk/aws-lambda/lib/lambda-version.ts#L137

What do you foresee the future of this API when this construct is available?

Reply:
Great question. The `edgeArn` API either just validates that the `$LATEST` version is not being used, or does some other minimal Edge compatibility checks. These latter checks would be the same that we'd want to incorporate into this construct (e.g., checking the runtime is supported). 

One option could be that the SSM `StringParameter` uses `currentVersion.edgeArn` instead of `currentVersion.functionArn`. That would mean the `us-east-1`-based function would call `edgeArn`, invoking the compatibility checks. 

This would limit the scope of this construct to only handling the cross-region piece, and `edgeArn` handling the rest of the validation. I think that might be the most consistent. Thoughts?

Reply:
My first thought is that we should consolidate and provide customers one (recommended, i.e., non-deprecated) way to create edge functions. 

How hard would it be to do everything you're doing as part of the `edgeArn()` API?

Reply:
Doing everything as part of the `edgeArn` API comes with a significant downside -- we'll end up with duplicate Functions.

```ts
const edgeArn = new lambda.Function(this, 'Fn', { /*...*/ }).edgeArn;
```

This would be a no-op if the current stack is in us-east-1, but if the stack is in any other region, we'd then need to create the Stack, duplicate Function (in us-east-1), and SSM parameter reader. This would result in two copies of the function -- one in the original region and one in us-east-1.  Lambda doesn't charge for at-rest functions (without provisioned concurrency) AFAIK -- only invocations and data transfer, so it shouldn't cost customers, but it's still I think fairly sub-optimal.

I can't think of a way to do everything here as part of `edgeArn` without duplicates, and given that, don't think it's the right path; that being said, open to disagreements.

I do agree on having a single way to do this; maybe one point of clarity would be to rename this construct `CrossRegionEdgeFunction`? Then it's clearer this is a wrapper for cross-region support, and not a general alternative to `Function`?

PS - I really don't _love_ `edgeArn` as the name for that function, as it only ever returns `this.functionArn`, potentially after doing validation; it implies that there is a scenario in which `edgeArn !== functionArn`, which is never the case. I seem to recall I was included on that PR and didn't speak up at the time, so I guess I can't complain.

Reply:
> I can't think of a way to do everything here as part of `edgeArn` without duplicates

Not sure if it's the right way or an elegant way, but this is a way to avoid duplicates in `edgeArn`:
```ts
public get edgeArn(): string {
  // ...

  const usEast1Condition = new CfnCondition(this, 'UsEast1Condition', {
    expression: Fn.conditionEquals(Stack.of(this).region, 'us-east-1'),
  });
  (this.lambda.node.defaultChild as CfnFunction).cfnOptions.condition = usEast1Condition ;

  const notUsEast1Condition = new CfnCondition(this, 'NotUsEast1Condition', {
    expression: Fn.conditionNot(Fn.conditionEquals(Stack.of(this).region, 'us-east-1')),
  });
  // cross region construct with SSM parameter using `notUsEast1Condition` here  
}
```

Reply:
Clever idea, @jogold.

The other disadvantage of doing this inline in `edgeArn` is that we'll still have to fix the circular dependency issue. The (current) solution relies on custom-resource, which itself has a dependency on lambda.

Reply:
https://github.com/aws/aws-cdk/tree/master/packages/%40aws-cdk/core#the-corecustomresourceprovider-class is a way to solve the circular dependency issue but of course it means partially rewriting something that is already written in `AwsCustomResource`.

Reply:
> `lambda.Version` already has an `edgeArn()` API that lets customers create an edge function from `Function` construct. What do you foresee the future of this API when this construct is available?

Coming back to the original question, I actually think this premise is a bit flawed: "... `edgeArn()` API that lets customers create an edge function from `Function`...". I don't think this is actually an accurate summary of the current state. `edgeArn` is a pseudo-internal API only used as an implementation detail by the CloudFront constructs (or at the very most, I suppose customers who are specifically building custom CloudFront constructs). For most customers, the input to the CloudFront APIs as still an `IVersion`, and they never call `edgeArn` explicitly.

For that reason, I don't think introducing `EdgeFunction` is introducing a second pathway here for customers. The current two paths for setting up Lambda@Edge are:

1. CloudFront stack is in `us-east-1` - Create/import `Function`, pass in `fn.currentVersion` as part of the CloudFront props.
2. CloudFront stack anywhere else - Create dedicated stack for new `Function` in `us-east-1`, use hard-coding, SSM parameters, or some other heavyweight solution to get the version ARN to the CloudFront stack. Eventually pass in `fnVersion` to the CloudFront props.

With the introduction of `EdgeFunction`, the path is:

1. User has an existing `Function` in us-east-1 - Import existing function, pass in `fn.currentVersion` as part of the CloudFront props.
2. No existing Lambda@Edge function - Create `EdgeFunction`, pass in `edgeFn` as part of the CloudFront props.

I think this is a pretty clear net win, and don't think the internals of the `edgeArn` API detract from that.

--- 

Completely orthogonal to this PR, I would be in favor of deprecating the `edgeArn` API entirely, and exposing the `checkEdgeCompatibility` API directly. That feels much more explicit about what the intent of the API is.